Rename the superpmi.py asmdiffs output directory

BruceForstall opened the pull request:
Previously, the directory name was `asm.os.arch.build_flavor`.
Now, use the MCH file name instead. Since the MCH filename of our
collections has the target os/arch/flavor, those components will still
be there (but will be based on the MCH file, not the host; this is
important if you are doing cross-compile asm diffs). It will
also have the MCH base name, such as "benchmarks.run", to help
distinguish different runs. Of course, if there is a conflict with
a previous run, `.1`, `.2`, etc., is appended to the directory name.

Also, add a new, optional `-tag` option to `superpmi.py asmdiffs` to
allow specifying an additional word to prefix the artifacts folder
base name. This might help more easily distinguish different experiments.

So, `superpmi.py asmdiffs -filter benchmarks -tag test2` might give you
dasm files in these directories:
```
c:\runtime\artifacts\spmi\asm.test2.benchmarks.run.windows.x64.checked.3\base
c:\runtime\artifacts\spmi\asm.test2.benchmarks.run.windows.x64.checked.3\diff
```

BruceForstall committed:
Rename the superpmi.py asmdiffs output directory

Previously, the directory name was `asm.os.arch.build_flavor`.
Now, use the MCH file name instead. Since the MCH filename of our
collections has the target os/arch/flavor, those components will still
be there (but will be based on the MCH file, not the host; this is
important if you are doing cross-compile asm diffs). It will
also have the MCH base name, such as "benchmarks.run", to help
distinguish different runs. Of course, if there is a conflict with
a previous run, `.1`, `.2`, etc., is appended to the directory name.

Also, add a new, optional `-tag` option to `superpmi.py asmdiffs` to
allow specifying an additional word to prefix the artifacts folder
base name. This might help more easily distinguish different experiments.

So, `superpmi.py asmdiffs -filter benchmarks -tag test2` might give you
dasm files in these directories:
```
c:\runtime\artifacts\spmi\asm.test2.benchmarks.run.windows.x64.checked.3\base
c:\runtime\artifacts\spmi\asm.test2.benchmarks.run.windows.x64.checked.3\diff
```

BruceForstall commented:
@dotnet/jit-contrib PTAL

kunalspathak commented on src/coreclr/scripts/superpmi.py:
```diff
@@ -1623,7 +1623,13 @@ def replay(self):
+                    # Use the MCH file base name as the main part of the directory for generated repro artifacts.
+                    # Remove the trailing ".mch", if any.
+                    artifacts_base_name = os.path.basename(mch_file)
+                    artifacts_base_name = artifacts_base_name[:-4] if artifacts_base_name.lower().endswith(".mch") else artifacts_base_name
```
May be repurpose `-tag` for these files as well? It might turn out useful.

BruceForstall replied:
It seems unlikely someone would specify `-tag` for a non-asmdiffs replay just to put the repro mcs in a differently named directory. I never even use the repro MC files, myself. I'm not sure I want complicate (further) the options to support that.

kunalspathak replied:
sounds reasonable.

kunalspathak commented on src/coreclr/scripts/superpmi.py:
```diff
@@ -1824,13 +1830,19 @@ def replay_with_asm_diffs(self):
+                # Remove the trailing ".mch", and prepend the tag, if any.
+                artifacts_base_name = os.path.basename(mch_file)
+                artifacts_base_name = artifacts_base_name[:-4] if artifacts_base_name.lower().endswith(".mch") else artifacts_base_name
+                artifacts_base_name = "{}.{}".format(self.coreclr_args.tag, artifacts_base_name) if self.coreclr_args.tag is not None else artifacts_base_name
```
If we decide to use `-tag` for saving replay failed files, then I would suggest to extract populating the `artifacts_base_name` in a function.

kunalspathak commented on src/coreclr/scripts/superpmi.py:
```diff
@@ -1907,12 +1919,12 @@ async def create_one_artifact(jit_path: str, location: str, flags: list[str]) ->
                         diff_items.append(item)
 
-                    logging.info("Creating dasm files")
+                    logging.info("Creating dasm files: %s %s", base_asm_location, diff_asm_location)
```
As we discussed offline, also display the exact `jit-analyze` command line to invoke?

kunalspathak replied:
Optional, if running asmdiff for private store, then pass `-count 0` to the `jit-analyze` command line so it will just show the summary instead of other details?

BruceForstall replied:
There's no private store yet in the script (that change isn't committed). I'm not in favor of this, since you'll end up manually running jit-analyze to see the data you want to see, anyway.

kunalspathak reviewed (commented):
Few comments.

BruceForstall committed:
Review feedback

Created `create_artifacts_base_name()` function.

Fixed `run_and_log` to display command line of command being
invoked at the specified log level. Currently only affects invoking
`jit-analyze`, which makes the `jit-analyze` command line visible
on the console, not just in the log file.

BruceForstall commented:
Change updated with feedback.

kunalspathak reviewed (approved):
LGTM. Thanks for doing this!
